src/injected-js/gmail/sync-compose-processor-20220909.ts

```diff
@@ -47,221 +30,288 @@ export function replaceBodyContentInComposeSendRequestBody_2022_09_09(
+function parseCreateUpdateSendDraftRequestBody(request: any[]) {
+  const updateList = request[1]?.[0];
 
-  if (thread) {
-    const threadId = parseThreadId(thread[0]);
-    if (!threadId) {
-      // exit cuz cannot parse
-      return null;
-    }
-
-    const msg =
-      thread[1]?.[2]?.[0]?.[4]?.[0] ||
-      /* Reply draft creation */ thread[1]?.[1]?.[0];
+  if (!Array.isArray(updateList)) {
+    // exit cuz cannot parse
+    return null;
+  }
 
-    if (!Array.isArray(msg)) {
-      // exit cuz cannot parse
-      return null;
-    }
+  const parsedMessages = updateList.map(parseRequestThread).filter(isNotNil);
 
-    const messageId: string = msg[0];
+  const sorted = sortBy(parsedMessages, (m) =>
+    ACTION_TYPE_PRIORITY_RANK.indexOf(m.type)
+  );
 
-    if (!messageId.startsWith('msg-')) {
-      // exit cuz cannot parse
-      return null;
-    }
+  return sorted[0] || null;
+}
 
-    const subject = msg[7];
-    const body = msg[8][1][0][1];
-    const actions = msg[10];
+/**
+ * Parses response body when compose window either saved a draft for the first time,
+ * updated the draft, or sent the draft.
+ * NOTE: response could contain multiple threads and messages within it
+ * even not related to a message/thread in the request body.
+ * So the calling code should find needed message manually.
+ */
+function parseCreateUpdateSendDraftResponseBody(response: any[]) {
+  const updateList = response[1]?.[5];
 
-    const hasRequiredActions =
-      intersection(actions, DRAFT_SAVING_ACTIONS).length ===
-      DRAFT_SAVING_ACTIONS.length;
+  if (!Array.isArray(updateList)) {
+    // exit cuz cannot parse
+    return [];
+  }
 
-    if (!hasRequiredActions) {
-      // exit if doesn't have required actions
-      return null;
-    }
+  return updateList
+    .map(parseResponseThread)
+    .filter(isNotNil)
+    .flatMap((parsedThread) => {
+      const { threadId, oldThreadId, parsedMessages } = parsedThread;
+
+      return parsedMessages.map((parsedMessage) => {
+        const { messageId, to, cc, bcc, actions, rfcID, oldMessageId } =
+          parsedMessage;
+
+        const actionType = actionsToComposeRequestType(actions);
+
+        if (!actionType) {
+          // unsupported actions within a message
+          return null;
+        }
+
+        return {
+          threadId,
+          messageId,
+          to,
+          cc,
+          bcc,
+          actions,
+          rfcID,
+          oldMessageId,
+          oldThreadId,
+          type: actionType,
+        };
+      });
+    })
+    .filter(isNotNil);
+}
 
-    return {
-      threadId,
-      messageId,
-      subject,
-      body,
-      actions,
-      type: 'FIRST_DRAFT_SAVE' as ComposeRequestType,
-    };
+function replaceBodyContentInSendRequestBody(
+  request: Array<any>,
+  newBodyHtmlContent: string
+) {
+  // since draftID is not passed from outside,
+  // use parse method to find a message which body needs to be replaced
```

Comment:
```suggestion
  // use parse function to find a message which body needs to be replaced
```

Methods are bound to classes.

Reply:
changed https://github.com/InboxSDK/InboxSDK/pull/825/commits/49251e3adf408d750c28b514896be270a12f58cc